lib/IDE/CodeCompletionConsumer.cpp

```diff
@@ -45,44 +59,23 @@ static MutableArrayRef<CodeCompletionResult *> copyCodeCompletionResults(
       case CodeCompletionDeclKind::AssociatedType:
       case CodeCompletionDeclKind::GenericTypeParam:
         return true;
-      case CodeCompletionDeclKind::PrecedenceGroup:
-      case CodeCompletionDeclKind::EnumElement:
-      case CodeCompletionDeclKind::Constructor:
-      case CodeCompletionDeclKind::Destructor:
-      case CodeCompletionDeclKind::Subscript:
-      case CodeCompletionDeclKind::StaticMethod:
-      case CodeCompletionDeclKind::InstanceMethod:
-      case CodeCompletionDeclKind::PrefixOperatorFunction:
-      case CodeCompletionDeclKind::PostfixOperatorFunction:
-      case CodeCompletionDeclKind::InfixOperatorFunction:
-      case CodeCompletionDeclKind::FreeFunction:
-      case CodeCompletionDeclKind::StaticVar:
-      case CodeCompletionDeclKind::InstanceVar:
-      case CodeCompletionDeclKind::LocalVar:
-      case CodeCompletionDeclKind::GlobalVar:
-      case CodeCompletionDeclKind::Macro:
-        return false;
+      default:
+        break;
       }
+    }
 
-      llvm_unreachable("Unhandled CodeCompletionDeclKind in switch.");
-    };
-  } else if (onlyPrecedenceGroups) {
-    shouldIncludeResult = [](const ContextFreeCodeCompletionResult *R) -> bool {
-      return R->getAssociatedDeclKind() ==
-      CodeCompletionDeclKind::PrecedenceGroup;
-    };
-  } else if (onlyMacros) {
-    shouldIncludeResult = [](const ContextFreeCodeCompletionResult *R) -> bool {
-      return R->getAssociatedDeclKind() ==
-      CodeCompletionDeclKind::Macro;
-    };
-  } else {
-    shouldIncludeResult = [](const ContextFreeCodeCompletionResult *R) -> bool {
-      // PrecedenceGroups are only valid in 'onlyPrecedenceGroups'.
-      return R->getAssociatedDeclKind() !=
-             CodeCompletionDeclKind::PrecedenceGroup;
-    };
-  }
+    if (R->getAssociatedDeclKind() == CodeCompletionDeclKind::PrecedenceGroup &&
+        Filter.contains(CodeCompletionFilterFlag::PrecedenceGroup))
+      return true;
+
+    if (R->getAssociatedDeclKind() == CodeCompletionDeclKind::Macro) {
+      CodeCompletionMacroRoles expectedRoles = getCompletionMacroRoles(Filter);
```

Comment:
could we hoist this out of `shouldIncludeResult`? `getCompletionMacroRoles(Filter)` is not particulary a heavy operation, but still.